### ISSUE TITLE
Nautical rendering style: 1. Add display of light characteristics in the nautical map 2. Add display of sectorial lights

### ISSUE THREAD

**Issue by josail** [osmandapp/OsmAnd#16894]
# 🚀 feature request

Nautical rendering style: 1. Add display of light characteristics in the nautical map 2. Add display of sctorial lights

### Description

1.
The display of the nautical rendering style lacks the important information on the light characteristics, which is essential for navigation by the light information. It is readily available in a well structured list of seamark tags as described in :
https://wiki.openstreetmap.org/wiki/Seamarks/Lights 
which can be evaluated in a appropriate text as in the openseamap.

2. 
A special display with lines and coloured arcs is required for sectorial lights, as e.g. displayed here: 
https://wiki.openstreetmap.org/wiki/Seamarks/Sectored_and_Directional_Lights

For 1. and 2. I'm willing to support with limited/beginners experience in the rendering style syntax. Thus the success will require engagement of a second person with more experience on the rendering style programming. I'm much willing to support by feedback on the openseamk syntax (https://wiki.openstreetmap.org/wiki/Seamarks/Light) and intended style (similar to openseamap and international standard INT-1 for nautical charts, https://wiki.openstreetmap.org/wiki/Seamarks/INT-1_Section_P).  And by repeated testing of adjustments to the nautical rendering style on an iOS iphone. 

### Describe the solution you'd like

1. Text display right of or below the navigational light in the nautical rendering style reflecting INT-1 nomenclature of the light characteristics (e.g. https://wiki.openstreetmap.org/wiki/Seamarks/INT-1_Section_P, www.openseamap.org)

2. lines, dashed lines and coloured arcs of sectorial lights around the navigational light like in  https://wiki.openstreetmap.org/wiki/Seamarks/Sectored_and_Directional_Lights

1. set up a composed string for text display in the nautical rendering style, some depence on zoom level according to major, minor and other navigational lights rating of importance

2. draw lines and arcs, graphical features to be implemented, I don't know how to do this

regarding 2.: alternatively as a first step to allow for the information of sectorial navigational lights as a workaround: Option for the display of a short 2nd text indicating the light sectors definitions

**Comment by sonora** [comment]
Let me cross-link to #16341. It's a frequent request from the boating community, so really quite a bit more than "nice to have" for the practical usability, is my understanding.

**Comment by josail** [comment]
Yes, I can underline that the seamark light characteristics are essential for nautical chart usability as aids to navigation at night. The flashing time, duration and multiple repetitions are unique and characteristic to clearly identify a seamark and its position in an area with certainty. Currently only the colour is indicated in the OSMAND nautical rendering style. This is insufficient, if you have e.g. three white cardinal lights around a point of danger. Only the flashing characteristic will tell, whether you need to go south or north of a specific light (seamark). So it is essential, and without light characteristics, it's of very limited use for navigation really.

**Comment by josail** [comment]
A starting point would be a strategy for the rendering types and its seamark section in lines 6742 to 7107:

**Comment by josail** [comment]
https://github.com/osmandapp/OsmAnd-resources/pull/965
addresses part 1 of the issue as far as possible today

**Comment by quantenschaum** [comment]
I came here from https://github.com/osmandapp/OsmAnd/issues/16342#issuecomment-1721330495

Sector lights can be displayed by adding dedicated lines to the OBFs for the sector limits and arcs. I wrote a [script](https://github.com/quantenschaum/mapping/blob/master/lightsectors.py) that does this and adjusted the [render style](https://github.com/quantenschaum/mapping/blob/master/marine.render.xml) accordingly.

Grab the OBF and style from https://github.com/quantenschaum/mapping/releases/tag/v3

Using these files it can look like this

![Screenshot_20230918-204835](https://github.com/osmandapp/OsmAnd/assets/3985599/bbb2bfee-f163-467e-a6b0-4e26beb1e849)

**Comment by quantenschaum** [comment]
I would like to propose a change of the light characteristics string. Currently the string does not follow the usual convention, it is similar but not the same. The light characteristics string should look like as described in the [INT1 section P, page 82](https://www.nauticalcharts.noaa.gov/publications/docs/us-chart-1/ChartNo1.pdf).

![image](https://github.com/osmandapp/OsmAnd/assets/3985599/05813602-cf76-4891-a54c-80ec4bdec8b8)

Main things to change

- group in round brackets `( )`
- list all colors, if light has multiple colors
- period with unit `s` seconds
- elevation with unit `m` meters
- range(s) with unit `M` nautical miles
- entire string w/o brackets but below the name of the light

**Comment by vshcherb** [comment]
- Label orientation and position is a problem we used to have textDy but I don't think w-e have it any more... One solution would be to have 2 lines and 1 transparent to display text over it bad solution anyway. 
- I'm not against changing text but I'm not sure how and where it needs to be done?

**Comment by vshcherb** [comment]
Mostly for simple text processing we use rendering_types.xml and only complex we hard code in java (less flexible and not transparent)

**Comment by quantenschaum** [comment]
Line breaks in a label are not possible?

**Comment by quantenschaum** [comment]
It seems to be like this: The text in `nameTag` and `nameTag2` are joined into a single string `nameTag (nameTag2)` that gets wrappend if it exceeds `textWrapWidth`. The wrapping occurs at space, period, comma, .... Non-breaking spaces (unicode 00A0) are not wrapped. This can be used to force the light characteristics onto the second line and not to break them in the middle.

**Comment by vshcherb** [comment]
Yes you are right ```textWrapWidth``` controls number of charactersl.

**Comment by josail** [comment]
I can confirm that "nameTag (nameTag2)" is what we could achieve June 2023, when I tried to imitate and allow display of light characteristics as far as possible, also by selection of different lightDetail in the app. 

Nevertheless, this is still far from replicating IHO Int-1, which should be the aim for next steps regarding both:

1. more complete text info on seamark lights 

2. sector light graphicsl display in colour arcs

My hint for a development in Java required for this certainly is: 

The OpenSeaMap and its rendering in OpenNavigationalCharts (ONC) is already able to provide text labels very close to Int-1 and draw arcs according to this standard. A lot of this rendering code is on Github and Sourceforge. 

I have had personal contact to the person achieving the Int-1 like display from OSM seamark tags in the ONC rendering. This is very likely also based on java code. I am personal not the right person to continue work on this. But I could forward an e-mail address, if you are interested to get in touch directly for next steps in OsmAnd. Please let me know. 

Regards Jo

**Comment by quantenschaum** [comment]
@josail What is "OpenNavigationalCharts"? Is there a link to the project/code?

It can be achieved as described in https://github.com/osmandapp/OsmAnd/issues/16894#issuecomment-1724202350. Either these extra lines and labels are generated during the creation of the OBFs or this data is generated separately and can be downloaded as extra OBFs by uses who need these features like it is the case with the depth data.

**Comment by josail** [comment]
Regarding the two options: 
Generation with the OBF in the app, i.e. rendering in the app would be best and allow better for up-to-date nautical chart display, much preferable in my view as active sailor

Several month delay of seamark changes would be a big disadvantage for OsmAnd. The OpenSeaMap and ONC do the rendering of seamark changes within few hours, sometimes even minutes. If a new download could lead to such up-to-date seamark light information then OsmAnd becomes a very valuable Offline source of most recent seamark and light information in the hands of sailors, more up-to-date than paper charts in many and oerhaps some very relevant cases.

**Comment by quantenschaum** [comment]
Yes, if this would happen in the app, this would be the best solution IMHO. But this is currently not possible, whereas generated extra lines in the OBFs are doable, as I have demonstrated.

Concerning updates, OsmAnd does not make OSM data accessible directly (You could use OSM raster tiles, though). It relies on the OBFs that are generated from the OSM data and which are updated monthly. So, generated light sectors in the OBFs could be updated monthly as well. I do not see a problem here. BTW: The ONC tiles are currently *not* updated!

And generally OSM data is not suitable for navigation anyways without comparing it thoroughly to other sources. I think it can be used for navigation, but you really have to know about the state of the data and data sources.

**Comment by quantenschaum** [comment]
I adjusted [my script](https://github.com/quantenschaum/mapping/blob/master/LIGHTS.md) such that a light characteristics string is generated and displayed in favor of the standard string. The characteristics of each individual sector are given in the sector arc. This is actually quite usable.

![image](https://github.com/osmandapp/OsmAnd/assets/3985599/6b2fb501-b73c-4505-b628-670b824bf53e)

**Comment by josail** [comment]
I perfectly agree. 

Good to learn that ONC is not up to date currently. Then something went out of work. Can you confirm at your example check that the OpenSeaMap is updated better, or the same update problem applies? I would then inform the group managing this accordingly, at best based on your OSM seamark data example.

**Comment by josail** [comment]
Thanks for the Schleimünde example from your updated script. This is a big improvement! I like to motivate you to implement this for  use of all in the OsmAnd. I would love to use this and prefer it much over my "symplistic" approach of June. The intuitive understandability of your Int-1 like display will be very helpful.

**Comment by quantenschaum** [comment]
I am not part of the OsmAnd team, but they are free to use the code from my repository. Would be nice to see this to become part of the OsmAnd nautical charts plugin. [But you can already use it](https://github.com/quantenschaum/mapping/blob/master/LIGHTS.md).

Does not belong here, but because you asked: AFAIK the services 

- https://map.openseamap.org/ nautical tiles: `https://tiles.openseamap.org/seamark/z/x/y.png`
- http://www.opennauticalchart.org/ nautical tiles: `https://t1.openseamap.org/seamark/z/x/y.png`

are essentially the same. Both show map tiles from `openseamap.org` and they are not up to date.

![Bildschirmfoto vom 2023-09-21 09-46-13](https://github.com/osmandapp/OsmAnd/assets/3985599/9804111b-83be-4154-a39a-cb2df858b609)

**Comment by quantenschaum** [comment]
Is there a keyword in the `render.xml` for scaling the icons? Or is there a path in the osmand directory where I can place custom icons and "shaders"?

**Comment by josail** [comment]
To my knowledge there are two folders



2. actual source for rendering:
Osmand
rendering_styles/style-icons

**Comment by quantenschaum** [comment]
Thanks. But there icons are inside the APK. Is there a location outside the APK, like in the osmand dir on the device, where I can place additional icons (w/o need to recompile)?

**Comment by josail** [comment]
I think the style-icons data from 2. is taken to the osmand APK or to some hidden dir. At least on my iOS device, which has limited file managment access, I do not have a visible seperate directory in the OsmAnd dir containing the style-icons. So, if it is there, it is hidden. I think there is a chance, that on Android devices, more access to the file structrure is granted (?).

**Comment by josail** [comment]
Here are my notes for managing and creating new icons: 
https://github.com/osmandapp/OsmAnd-resources/blob/master/icons/tools/Generating_OSMAND_map_icons.txt

**Comment by josail** [comment]
@vshcherb and @sonora : 
Would someone more deeply involved in the java development of OSMAND be able to make good use of the code suggested by quantenschaum ? 

> I am not part of the OsmAnd team, but they are free to use the code from my repository. Would be nice to see this to become part of the OsmAnd nautical charts plugin. [But you can already use it](https://github.com/quantenschaum/mapping/blob/master/LIGHTS.md).

It sounds like a perfect "winter project" and likely quantenschaum and certainly myself would be happy testers of such new features. 

The rendering of arcs from nautical sectorial light information in offline vector OSM data would be an outstanding and unique feature to my knowledge. Up to now this is only availabel in [OpenNautiCalcharts](http://www.opennauticalchart.org/?permalink=true&zoom=17&lat=54.67153&lon=10.03752&seamarks=true&coordinate_grid=true) and [OpenSeaMap](https://map.openseamap.org/?zoom=17.2&lon=10.03611&lat=54.67139&layers=TFTFFFTFFTFFFFFFTFFFFF). Both of these can be used on mobile devices only with online connection (which is often missing on the water) or as raster tile download (very data intensive and laboursome for large areas of potential navigational approaches). So there is clearly added value and request for such a new feature in OSMAND, as pointed out by sonora: 

> Let me cross-link to #16341. It's a frequent request from the boating community, so really quite a bit more than "nice to have" for the practical usability, is my understanding.

**Comment by vshcherb** [comment]
Yes we monitor @quantenschaum his work progress and so far he is doing pretty well. First of all most of his work is processing contour lines which is really hard work in QGis. Rendering arcs as scripts was my idea as well and it can't be improved with java, we can't render such complicated arcs in rendering engine. 
If @quantenschaum continue his work, we will have no option as to switch to his work as primary source of data and make nautical plugin. 

I feel more & more confident that work is on the right track and today we won't have even good expertise to replicate this work

**Comment by vshcherb** [comment]
@quantenschaum we're providing Nautical seamarks already as a separate file and we will keep doing it. So in that case indeed we need to merge as much rendering style features as possible into nautical.render.xml and think about files distribution. 

Files could be distributed by our servers if you can generate it monthly and  give us permission to distribute it.

**Comment by quantenschaum** [comment]
@vshcherb You mean the world seamarks? Seamarks are also included in the regular OBFs for each country/region. This means, that if you download an OBF for a country and the world seamarks, you have duplicate data in OsmAnd. If the data in both is the same, it does not matter, but if not (live updates!), objects might appear twice. OK, but this just as a thought. You do not want to replace any of the existing OBFs, just add another one that optionally can be downloaded. There are about 37500 lights worldwide with a `range`. They can be put into a single OBF, I think.

**Comment by vshcherb** [comment]
Yes it should be alternative file I believe cause combination make things difficult. So user should install either your version of obf or standard seamarks

**Comment by vshcherb** [comment]
You can use https://creator.osmand.net/osm-extract/World_seamarks/ our pbf to generate file with more features built-in

**Comment by quantenschaum** [comment]
Well, IMHO it is simpler to provide an additional file that contains the light sectors only. I prefer to used the individual country OBFs instead of the world seamarks, because I also want an accurate coastline and features on land. So it is desirable to have the light sectors only in a separate file to be able to use them in conjunction with whatever other OBFs. Since the file with worldwide sectors is only ~15MB, it is not worth splitting it into smaller regions.

**Comment by quantenschaum** [comment]
I set up a cronjob that generates the sectors at least once a week.

http://waddenzee.duckdns.org/Lightsectors.obf

**Comment by josail** [comment]
The solution offered by quantenschaum is convincing. I have tested it on iOS iphone 14 with following OsmAnd result: 
![IMG_0629](https://github.com/osmandapp/OsmAnd/assets/93162192/a9d65f59-3cd1-424a-b003-b14e20654576)
![IMG_0628](https://github.com/osmandapp/OsmAnd/assets/93162192/01adddcc-3e43-481b-a09f-5587838e2bd1)

**Comment by quantenschaum** [comment]
Could anyone make use of the light sectors in https://github.com/osmandapp/OsmAnd/issues/16894#issuecomment-1780749215 ?

**Comment by josail** [comment]
@vshcherb and others
quantenschaum and I lack the abilities of putting his solution into work in OsmAnd, but we are both longing very much for it. Could you or some other person of the team put some time on making this important step for the boating community?

As demonstrated step by step manually, all elements described by quantenschaum work perfectly. However, I would not know, how to implement that professionally (not manually, and sufficient in the OsmAnd world, i.e. independent of updating and downloading from a file provided on a server by quantenschaum), so that it becomes available to the whole community. 

More correct seamark light display would certainly enhance the usability of the OsmAnd Nautical render map for boaters significantly!

a) I think the basis would be to enhance the creation of the .obf data taking the path provided in the tools of quantenschaum.

b) When it comes to integrating quantenschaums [marine rendering style](https://github.com/quantenschaum/mapping/blob/master/marine.render.xml) with OsmAnd's nautical rendering style:

- I see no significant conflicts (if any) with the existing nautical rendering style, but rather important amendment for more complete display of light characteristics. It would really advance this issue from my preliminary solution (1.,  text only, still partially incomplete and deviating from nautical chart standards) to the intended better level (2., light arcs and complete text for light characteristics, see first comment of this issue.) 

- If there is work for consolidating the nautical rendering style on this 2. level, I guess both of us would be very much willing to contribute to this final step. But we need help for the advanced level of work on step a). Who can support this?

**Comment by vshcherb** [comment]
Hi @josail we certainly need to take ownership of it. I do appreciate the work you have done and I see that we need to integrate marine style replace our default boat rendering style. Also we need to integrate the data and schedule it on our side to produce monthly with World_seamarks.obf 

I have 1 question and 1 concern:
- Are you fine to merge your work into OsmAnd repository ?
- Last time we've tried to follow the steps you've done we've realized it's quite a lot of work done manually and we had a concern whether we could do the same steps globally. Especially in some regions where the data was messy. 

We're certainly willing to take leadership and automate it further. We don't know how to approach it yet (transfer maintenance is not easy unfortunately)

**Comment by josail** [comment]
Hi Vshcherb, thanks for your engaging reply. The additional work proposed for the 2nd step of improving seamark lights originates from @quantenschaum. Thus I like to forward your question and concern to him/her?

- for the question I read the quntenschaums contributions in this issue as a encouraging "yes", but better to be confirmed by the original author
- The assessment of how much has to be done manually, effort of maintainance is also better assessed by quantenshaum. 

Should the proposed solution really be related to a lot of manual effort, then I can suggest to also consult with @malcolmh, who has implemented java solutions for highly automate, global and "within minutes" OSM data seamark rendering with very similiar result, i.e. including light arcs and INT-1 standard text display next to lights, both for the opennauticalchart and openseamap here: https://github.com/OpenNauticalChart/renderer, see also some more description at https://wiki.opennauticalchart.org/

see the results corresponding to the tree examples illustrated above in this issue: 
- Schleimünde: http://www.opennauticalchart.org/?permalink=true&zoom=16&lat=54.67172&lon=10.03759&seamarks=true&coordinate_grid=true
- Tegeler Plate: http://www.opennauticalchart.org/?permalink=true&zoom=15&lat=53.79801&lon=8.18977&seamarks=true&coordinate_grid=true
- Helgoland: http://www.opennauticalchart.org/?permalink=true&zoom=16&lat=54.17590&lon=7.90028&seamarks=true&coordinate_grid=true

**Comment by quantenschaum** [comment]
Yes, please use my work and integrate it into OsmAnd if you like.

data created via cronjob is available at http://waddenzee.duckdns.org/download/

**Comment by quantenschaum** [comment]
Hi, is there any progress on this topic?

It would be nice get my [marine rendering style](https://github.com/quantenschaum/mapping/blob/master/marine.render.xml) merged (could replace the [nautical style](https://github.com/osmandapp/OsmAnd-resources/blob/master/rendering_styles/nautical.render.xml)) It contains many changes/enhancement to make the chart much more usable as a nautical chart, but is not that easy to merge and it get's worse the more they diverge. I also integrated the [nautical depth style](https://github.com/quantenschaum/mapping/blob/master/depthcontourlines.addon.render.xml) into it, because nautical depth are an integral part of a nautical chart (and currently the depth style gets overwritten with every update of OsmAnd).

There also needs to be some additional data (i.e. light sectors, depth contours and areas, depth expositions, etc.) in the OBFs themselves, so the [rendering types](https://github.com/quantenschaum/mapping/blob/master/rendering_types.xml) have to be updated as well.

![hiddensee](https://github.com/osmandapp/OsmAnd/assets/3985599/a012e8d7-fed8-48ab-ac0e-3ea460d94c4e)

![norderney](https://github.com/osmandapp/OsmAnd/assets/3985599/aba29004-0b8b-4ada-89c6-2a06f7c77823)

**Comment by vshcherb** [comment]
@quantenschaum progress was interrupted by complicate 4.7 release. Just yesterday discussed that we need to do it this / next week at least start. 
We discussed the main problem here is to create or not create primitive to draw light sectors. So they will be visible on all scales equally but that's quite a task.

**Comment by quantenschaum** [comment]
Ok, nice. What do you mean by?

>the main problem here is to create or not create primitive to draw light sectors. So they will be visible on all scales equally but that's quite a task.

The marine style contains the styling for the light sectors but also contains other modifications, this could be merged now. The creation of the OBFs containing the light sector primitives is a separate task. One could start with a single set of light sectors as in the `lightsectors.obf` that I provide. This is already very usable. At a later stage one could generate an OBF that contains that contains multiple sets of lightsectors for different zoom level.

**Comment by quantenschaum** [comment]
Using [my script](https://github.com/quantenschaum/mapping/blob/master/lightsectors.py) one can create lightsector of different sizes using the `-a` and `-f` options. Then these sectors are compiled into OBFs at diffrent zoom levels, the OBFs are combined into a single OBF and you get diffrently scaled lightsectors at different zoom levels. This way you can avoid to make any changes to render engine or to generate the geometry for the light sectors in the renderer on the fly.

I just [implement this](https://github.com/quantenschaum/mapping/blob/b28d7364c43a6d277a2f61493319485db1c624b5/makefile#L241) for 3 levels. It looks like this.


https://github.com/osmandapp/OsmAnd/assets/3985599/3384ff77-b234-4fd5-96ce-f50ad5afbfc8

**Comment by vshcherb** [comment]
- Light sector has a fixed size now in "meters" not in "pixels". So if you zoom you would see full inside the sector. Though we're going to proceed with it as the beginning.

**Comment by quantenschaum** [comment]
What exactly do you mean?
Yes, the sector arc is a line consisting of several points which have lat/lon positions, so they "do not know" anything about pixels on screen. By generating multiple arcs for different zoom levels, the arcs are always within the viewport. One can also set the range threshold such that low range lights are only displayed at higher zoom levels.

Is there a possibility to give the sector arc radius in pixels? That would require the arc to be generated by the renderer, in a geometry shader. That would be nice solution. Are geometry shaders available on all current mobile devices? Alternative the arcs have to generate in demand in software (not opengl).

**Comment by quantenschaum** [comment]
I just updated the lightsectors file with now 4 different levels with different arc radii and thresholds to include lights based on their range. Try it, I think, this is quite usable.

http://waddenzee.duckdns.org/download/

**Comment by josail** [comment]
quantenschaum, the nautical chart displays on your website look really promising. That's a whole new level of 
1st accurate nautical chart display and
2nd so much needed reliable depth information for Netherlands and Germany, where hydrographical offices depth information is openly accessible.
 
From some displays by Melcolm I saw that he suggested similar displays for UK coast. So I guess, there must be some open access depth information from UKHO, too? and Norway seemed pretty open with its data policy. 

I hope other countries hydrographic offices will follow up with open access bathemetry data e.g. in the framework of the international [(https://portal.opentopography.org/datasetMetadata?otCollectionID=OT.122023.4326.1*) GEBCO Project]. So far the GEBCO project seems good for open sea areas, but with 500m x 500 m resolution not jet good enough for coastal navigation and navigational obstructions.

I much hope we get the accassible information into a professional INT-1 like OSMAND display with nautical or marine rendering and the tools developed by quantenschaum.

**Comment by quantenschaum** [comment]
Hi @josail, some notes and questions

depth data sources I know of

- :netherlands: [contours and spot soundings available for Waddenzee and Zeeland](https://www.vaarweginformatie.nl/frp/main/#/page/infra_enc)
- :de:  [contours available for North Sea and Baltic Sea](https://www.geoseaportal.de/mapapps/resources/apps/navigation), but no spot soundings
- :norway:  @miramikes digged out [something](https://github.com/osmandapp/OsmAnd/discussions/18116#discussioncomment-7109414): [Sjøkart - Dybdedata](https://kartkatalog.geonorge.no/metadata/sjoekart-dybdedata/2751aacf-5472-4850-a208-3532a51c529a)
- :gb:  UKHO seems very restrictive, I did not find anything freely accessible, you can [buy data](https://datahub.admiralty.co.uk/portal/apps/sites/#/marine-data-portal)

Any coarse depth data and anything deeper than 10m are actually not important for coastal navigation (of pleasure craft). What we need is accurate and up to date depth data <10m. The only reliable sources seem to be the hydrographic offices.

The depth data discussion is here https://github.com/osmandapp/OsmAnd/discussions/18116

>From some displays by Melcolm I saw that he suggested similar displays for UK coast.

Where did you see that? Could you provide a link?

**Comment by RZR-UA** [comment]
Dev notes:

- [x] add support for the light sectors to Nautical style
- [x] import new Marine style, add localization strings to strings.xml
- [ ] create Jenkins job to generate and integrate sectors into World_seamarks
- [ ] port changes to OsmAnd iOS, test both styles on World_seamarks
- [ ] port lightsectors code from Python to Java with threads (later)

**Comment by RZR-UA** [comment]
@quantenschaum 

We are going to accept your superb Marine style into OsmAnd Android/iOS apps.

It looks great, is highly customizable, and prettier than classic Nautical style.

Once we adopt the new style, both styles will need to be maintained well.

The new file is now based on the reformatted nautical xml file.

It's hard to make a correct diff between new and original files.

Could you try to re-create marine.render.xml based on unchanged nautical.render.xml ?
It should not have any formatting such as splitting long lines and replacing tabs and spaces.

Alternatively, you could try to re-format marine.render.xml and then run `diff -u -w marine.render.xml nautical.render.xml` to verify that there are only your actual changed at the output.

**Comment by quantenschaum** [comment]
Nice! - I will try to get rid of the formatting.

I also thought about having the nautical and the marine style in parallel, but it seems to be a lot of tedious manual work to port all changes made to the nautical style to the marine style as well. This is what I did to keep to it in sync with your nautical style which it is based on. So, IMHO it is maybe best to have just a single style for nautical charts, and the marine style can be configured to look like the nautical style. This make maintenance much easier. 

I also merged the styling of the depth contours into the marine style (marked by comment `depthcontourslines`) to avoid them getting overwritten by updates. These could replace `depthcontourlines.addon.render.xml`, but this also would require the depth data to updated to the modified tagging scheme used in `rendering_types.xml`.

**Comment by xmd5a2** [comment]
@quantenschaum Ideally marine style should be dependent on it's base - nautical. I mean it is better to create a style that contains only necessary changes to nautical. Look at [offroad](https://github.com/osmandapp/OsmAnd-resources/blob/master/rendering_styles/offroad.render.xml) style.
`<renderingStyle name="Offroad by ZLZK" depends="topo" defaultColor="#ffffff" version="1">`
It is tiny but it uses topo style as a base. And usually there is no need to update offroad when topo is updated. All changes will work automatically.
I don't think that merging depth contours to style is a good idea. I think your changes should be integrated into our depthcontourlines.addon.render.xml somehow.

**Comment by quantenschaum** [comment]
OK, It didn't know that this was possible. How are the instructions of both styles merged, you can add new instructions, but can you also override existing instructions?

**Comment by xmd5a2** [comment]
Yes, absolutely. But if you need to change for example water color and there is no renderingAttribute / renderingConstant in nautical style then better to create renderingAttribute in nautical style than copying a big chunk of code to marine style.

**Comment by vshcherb** [comment]
1. https://creator.osmand.net:8080/view/OSM%20Planet/job/Planet_Update-Extract%20Osm%20maps%20(seamarks)/ Generate all needed for marine / nautical data in World_seamarks.xml 
2. Update OsmAndMapCreator to produce detailed (as needed) World_seamarks.obf
3. Integrate Marine style into OsmAnd styles collection

**Comment by vshcherb** [comment]
Work in progress 
1. Added new style marine in the OsmAnd Android
2. Published joined World_lightsectors_2.obf as World_seamarks_2.obf
3. New style marine in the OsmAnd iOS

To do:
- [ ] Update marine style to the latest version if needed / possible - not crashes @xmd5a2 
- [ ] Finalize monthly update maps process... @vshcherb 
- [ ] Check descriptions information @dmpr0 

Follow up tasks:
- Merge sectors into standard maps ? - not really possible due to live updates and aritificial objects
- [ ] Update parameter names descriptions

**Comment by vshcherb** [comment]
https://github.com/osmandapp/OsmAnd-resources/commit/3d904efd80d14bfc998f38af7fdb44ab6b4a09ca

**Comment by vshcherb** [comment]
new utility has added to generate target  (for now without parameters but they could be added later)
```
./OsmAndMapCreator/utilities.sh generate-lightsectors World_lightsectors_src.osm.gz .
```
It produces combined file of all lightsectors for all zooms World_lightsectors_src.obf. World_lightsectors_src.osm.gz  is osm file from overpass.

**Comment by vshcherb** [comment]
New maps are published for testing.
Some differences.
1. Java code doesn't use mercaator projection for size of arcs / distances and uses SCALE_RANGE_DIFF_PYTHON = 0.62, so it's more or less aligned with python code around 45 degree of latitude
2. LEADING_LIGHT_CATEGORIES - Arrays.asList("leading", "front", "rear", "lower", "upper")  these objects are ignored for directional lines cause in the code I see option that need to be ignored by default, however old map includes some of them probably it's a bug... like https://www.openstreetmap.org/node/1452252610#map=19/51.959501/4.145070

**Comment by Zirochkabila** [comment]
# Testing 1
`OsmAnd~ 5.2.0#5845m, Released: 2025-08-14`
`5.2.1.5211`


Checked the display of light characteristics on the nautical chart (map from 13.08.2025) (iOS & Android (V1 & V2)) - **Light characteristics are displayed.**

The screenshots show the light characteristics - “Fl R 2s” (flashing red, 2-second cycle) and “LFl G 6s” (long flashing green, 6-second cycle) are exactly that. They describe the color, frequency and type of flashing.

<details><summary>Details</summary>
<table style="width:100%">
  <tr>
<th><img src="https://github.com/user-attachments/assets/7fb8dac9-369f-4991-b6ab-335cfd9e7bbb"  width="300" height="auto"/> </th>
 <th><img src="https://github.com/user-attachments/assets/68bdd866-ce9b-4dca-8c5c-055f46fc1da5"  width="300" height="auto"/> </th>
  </tr>
</table>  
</details>

**Comment by dmpr0** [comment]
## UI Review 

<table style="width:100%">
  <tr>
    <th>iOS</th>
    <th>Android</th>
  </tr>
  <tr>
    <td><img src="https://github.com/user-attachments/assets/857e54b9-4fdb-437d-a8c9-7c872a78a515"  width="300" height="auto"/> </td>
    <td><img src="https://github.com/user-attachments/assets/bb98e641-e4ad-448d-bdb2-3e19a0df10b1"  width="300" height="auto"/> </td>
  </tr>
</table>   


- [ ] Switches doesn't work on iOS: `ENC markers`, ...
- [ ] Better to update or make some changes to all added functions in UI on Android too
